File: packages/swiftui/Sources/CocsoUI/CCSpinner.swift
import SwiftUI

/// An indeterminate activity indicator.
public struct CCSpinner: View {
    private let variant: CCSpinnerVariant
    private let size: CCSpinnerSize
    private let label: String

    @Environment(\.colorScheme) private var colorScheme
    // The brand the app set at its root. Tokens and resolvers take it so a
    // design-system view draws the same primary the app does.
    @Environment(\.cocsoBrand) private var brand
    @Environment(\.accessibilityReduceMotion) private var reduceMotion
    @State private var spinning = false

    public init(
        variant: CCSpinnerVariant = .primary,
        size: CCSpinnerSize = .medium,
        label: String = CCStrings.loading
    ) {
        self.variant = variant
        self.size = size
        self.label = label
    }

    public var body: some View {
        let style = CCSpinnerStyle.resolve(variant: variant, size: size, scheme: colorScheme, brand: brand)
        let side = style.output ?? 16
        let bladeCount = style.blades ?? 8
        ZStack {
            ForEach(0..<max(bladeCount, 1), id: \.self) { index in
                RoundedRectangle(cornerRadius: style.bladeRadius ?? 1)
                    .fill(style.bladeColor ?? CocsoTokens.Color.interactivePrimary(colorScheme, brand: brand))
                    .frame(width: style.bladeWidth ?? 2, height: style.bladeHeight ?? 5)
                    .offset(y: -side / 2 + (style.bladeHeight ?? 5) / 2)
                    .rotationEffect(.degrees(Double(index) / Double(max(bladeCount, 1)) * 360))
                    .opacity(Double(index + 1) / Double(max(bladeCount, 1)))
            }
        }
        .frame(width: side, height: side)
        // One turn per `duration-decorative`, the period of the web's
        // `spinner-blade-fade` — a rotating opacity ramp is the same picture as
        // eight blades fading in turn.
        .rotationEffect(.degrees(spinning && !reduceMotion ? 360 : 0))
        // Reduced motion: the web swaps the spin for a slow pulse of the whole
        // indicator (`spinner-reduced-pulse`, 0.4 ↔ 0.8 over 2s), and so does
        // this — still says work is happening, without anything travelling.
        .opacity(reduceMotion ? (spinning ? 0.8 : 0.4) : 1)
        .animation(
            reduceMotion
                ? CocsoTokens.Easing.default(1).repeatForever(autoreverses: true)
                : .linear(duration: CocsoTokens.Duration.decorative).repeatForever(autoreverses: false),
            value: spinning
        )
        .onAppear { spinning = true }
        .accessibilityElement()
        .accessibilityLabel(label)
    }
}

File: packages/compose/src/main/kotlin/ai/cocso/ui/CCSpinner.kt
package ai.cocso.ui

import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.drawscope.rotate
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp

/** An indeterminate activity indicator. */
@Composable
fun CCSpinner(
    modifier: Modifier = Modifier,
    variant: CCSpinnerVariant = CCSpinnerVariant.primary,
    size: CCSpinnerSize = CCSpinnerSize.medium,
    label: String = "Loading",
) {
    val style = cCSpinnerStyle(variant = variant, size = size)
    val side = style.output ?: 16.dp
    val bladeCount = (style.blades ?: 8).coerceAtLeast(1)
    val bladeWidth = style.bladeWidth ?: 2.dp
    val bladeHeight = style.bladeHeight ?: 5.dp
    val color = style.bladeColor ?: CocsoTokens.Color.interactivePrimary()
    val reduced = reducedMotion()

    // One turn per `duration-decorative`, the period of the web's
    // `spinner-blade-fade` — a rotating opacity ramp is the same picture as
    // eight blades fading in turn.
    val rotation = if (reduced) {
        0f
    } else {
        rememberInfiniteTransition(label = "spinner").animateFloat(
            initialValue = 0f,
            targetValue = 360f,
            animationSpec = infiniteRepeatable(
                tween(CocsoTokens.Duration.decorative, easing = LinearEasing)
            ),
            label = "spinner-rotation",
        ).value
    }
    // Reduced motion: the web swaps the spin for a slow pulse of the whole
    // indicator (`spinner-reduced-pulse`, 0.4 ↔ 0.8 over 2s), and so does this
    // — still says work is happening, without anything travelling.
    val pulse = if (reduced) {
        rememberInfiniteTransition(label = "spinner-pulse").animateFloat(
            initialValue = 0.4f,
            targetValue = 0.8f,
            animationSpec = infiniteRepeatable(
                tween(1000, easing = CocsoTokens.Easing.default),
                RepeatMode.Reverse,
            ),
            label = "spinner-opacity",
        ).value
    } else {
        1f
    }

    Canvas(
        modifier = modifier
            .size(side)
            .semantics { contentDescription = label }
    ) {
        val w = bladeWidth.toPx()
        val h = bladeHeight.toPx()
        val r = (style.bladeRadius ?: 1.dp).toPx()
        val corner = CornerRadius(r, r)
        repeat(bladeCount) { index ->
            val degrees = rotation + index.toFloat() / bladeCount * 360f
            rotate(degrees = degrees, pivot = center) {
                drawRoundRect(
                    color = color,
                    topLeft = Offset(center.x - w / 2f, 0f),
                    size = Size(w, h),
                    cornerRadius = corner,
                    alpha = (index + 1).toFloat() / bladeCount * pulse,
                )
            }
        }
    }
}
